### PR TITLE
terminal: Fix copy to clipboard lag

### LINKED ISSUE #7322: terminal: Events are delayed (e.g. copy to clipboard)
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

How I noticed:
When copying terminal output and then immediately pasting it again into the terminal I almost always get what I had in the clipboard before. If I wait a bit between the copy and the paste it works.

Now that Zed is open source 🥳 I investigated the issue myself and I think I know what is going on.
If I press Cmd+C to copy, all the terminal does is put a "copy request" into a queue
```rs
self.events.push_back(InternalEvent::Copy)
```
(`terminal.rs:912`).
Then the next time GPUI paints the terminal it calls `terminal.sync` which goes through mentioned queue and actually puts the currently selected terminal text into the system clipboard.
```rs
InternalEvent::Copy => {
    if let Some(txt) = term.selection_to_string() {
        cx.write_to_clipboard(ClipboardItem::new(txt))
    }
}
```
(`terminal.rs:713`)

It seems like paint is called with a fairly low frequency and so there is enough time for me to paste before the copy actually happenes. I also see other race conditions here, e.g. the selected text could theoretically change between the copy "request" and the actual copying into the system clipboard.

I feel like the render loop isn't the right place to handle such kind of events, but I actually have not much of an idea what I'm talking about so happy to be corrected.

I have a fix for copy to clipboard and will open a PR shortly.

### Environment

Zed: v0.121.2 (Zed Preview)
OS: macOS 14.2.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

### PR DESCRIPTION
**Pull request by N8th8n8el**
Fixes #7322

### CODE REVIEW

**Comment by N8th8n8el** [comment]
@cla-bot check

**Comment by mrnugget** [comment]
So is the idea here that you skip the event loop entirely? Is there no alternative solution possible where we wake up the event loop listener on certain events?

**Comment by N8th8n8el** [comment]
> So is the idea here that you skip the event loop entirely? Is there no alternative solution possible where we wake up the event loop listener on certain events?

If by event loop you mean paint loop (at least that is my current understanding) then yes.
As stated in the issue, I feel like waiting for (or triggering) paint to copy something into the clipboard doesn't make sense. Ifaics copying to the clipboard has no effect on what is rendered, so why do we need to (wait for) paint?

Maybe it makes more sense if we also take other events into account though, such as clearing the terminal which does have an effect on what is rendered.

**Comment by mrnugget** [comment]
> If by event loop you mean paint loop (at least that is my current understanding) then yes.
> As stated in the issue, I feel like waiting for (or triggering) paint to copy something into the clipboard doesn't make sense. Ifaics copying to the clipboard has no effect on what is rendered, so why do we need to (wait for) paint?

I don't think the event loop in the terminal is strictly about rendering/painting. It's about the state of the terminal itself.

And this bug (and others that have popped up) make me think that maybe something with the event loop and how often it runs is broken, not just with the copy behaviour itself.

**Comment by nathanael-ruf** [comment]
> And this bug (and others that have popped up) make me think that maybe something with the event loop and how often it runs is broken, not just with the copy behaviour itself.

Yes it's not just copying (that's
>I also believe that this isn't a general fix of the underlying issue

---

> I don't think the event loop in the terminal is strictly about rendering/painting. It's about the state of the terminal itself.

`terminal.sync(cx);` is called in `TerminalElement:compute_layout` which is called in `TerminalElement:paint`. I don't know yet how GPUI works, i.e. how it decides when to paint, but from the name it sounds like this is just about painting and e.g. copying to the clipboard shouldn't trigger a repaint (maybe?).

**Comment by N8th8n8el** [comment]
> And this bug (and others that have popped up) make me think that maybe something with the event loop and how often it runs is broken, not just with the copy behaviour itself.

Yes it's not just copying (that's
>I also believe that this isn't a general fix of the underlying issue

---

> I don't think the event loop in the terminal is strictly about rendering/painting. It's about the state of the terminal itself.

`terminal.sync(cx);` is called in `TerminalElement:compute_layout` which is called in `TerminalElement:paint`. I don't know yet how GPUI works, i.e. how it decides when to paint, but from the name it sounds like this is just about painting and e.g. copying to the clipboard shouldn't trigger a repaint (maybe?).

File: crates/terminal/src/terminal.rs [hunk 1]
```diff
@@ -710,11 +709,6 @@ impl Terminal {
                 }
             }
 
-            InternalEvent::Copy => {
```
Comment (mikayla-maki) [inline crates/terminal/src/terminal.rs:713]: I don't think we should remove this, as we'd like to be cooperative and give the PTY time to write into the Terminal.

That said, I think the better solution would be to...

File: crates/terminal_view/src/terminal_view.rs [hunk 2]
```diff
@@ -451,7 +451,9 @@ impl TerminalView {
 
     ///Attempt to paste the clipboard into the terminal
     fn copy(&mut self, _: &Copy, cx: &mut ViewContext<Self>) {
-        self.terminal.update(cx, |term, _| term.copy())
```
Comment (mikayla-maki) [inline crates/terminal_view/src/terminal_view.rs:454]: Add a `cx.notify()` here to schedule the re-render.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
The deeper solution to this problem IMO is to take ownership of the Alacritty event loop code and let it own the alacritty Terminal struct. Then we can send everything in the existing, poorly named, `EventLoop` struct to the Alacritty terminal as messages that can be replied to instantly. 

Lacking that refactoring though, I think we should stick to the current solution which maintains the relative ordering of events in time.

**Review by mikayla-maki — APPROVED** [review]

**Pull request merged**